components/motor/ulnstepper/28byj-48.go

```diff
@@ -0,0 +1,368 @@
+
+	if math.Signbit(revolutions) != math.Signbit(rpm) {
+		d = -1
+	}
+
+	revolutions = math.Abs(revolutions)
+	rpm = math.Abs(rpm) * float64(d)
+
+	targetPosition := m.stepPosition + int64(float64(d)*revolutions*float64(m.ticksPerRotation))
+
+	// stepperDelay is the wait time between each step taken.
+	// The minimum value is set to 0.002s, anything less then this can potentially damage the gears.
+	stepperDelay := time.Duration(int64((1/(math.Abs(rpm)*float64(m.ticksPerRotation)/60.0))*1000000)) * time.Microsecond
+	if stepperDelay < minDelayBetweenTicks {
+		m.logger.Debugf("Computed sleep time between ticks (%v) too short. Defaulting to %v", stepperDelay, minDelayBetweenTicks)
+		stepperDelay = minDelayBetweenTicks
+	}
+
+	return targetPosition, stepperDelay
+}
+
+// GoTo instructs the motor to go to a specific position (provided in revolutions from home/zero),
+// at a specific RPM. Regardless of the directionality of the RPM this function will move the motor
+// towards the specified target.
+func (m *uln28byj) GoTo(ctx context.Context, rpm, positionRevolutions float64, extra map[string]interface{}) error {
+	curPos, err := m.Position(ctx, extra)
+	if err != nil {
+		return errors.Wrapf(err, "error in GoTo from motor (%s)", m.motorName)
+	}
+	moveDistance := positionRevolutions - curPos
+
+	m.logger.Debugf("Moving %v ticks at %v rpm", moveDistance, rpm)
+
+	if moveDistance == 0 {
+		return nil
+	}
+
+	return m.GoFor(ctx, math.Abs(rpm), moveDistance, extra)
+}
+
+// Set the current position (+/- offset) to be the new zero (home) position.
+func (m *uln28byj) ResetZeroPosition(ctx context.Context, offset float64, extra map[string]interface{}) error {
+	return motor.NewResetZeroPositionUnsupportedError(m.motorName)
+}
+
+// SetPower is invalid for this motor.
+func (m *uln28byj) SetPower(ctx context.Context, powerPct float64, extra map[string]interface{}) error {
+	return errors.Errorf("raw power not supported in stepper motor (%s)", m.motorName)
+}
+
+// Position reports the current step position of the motor. If it's not supported, the returned
+// data is undefined.
+func (m *uln28byj) Position(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return float64(m.stepPosition) / float64(m.ticksPerRotation), nil
+}
+
+// Properties returns the status of whether the motor supports certain optional features.
+func (m *uln28byj) Properties(ctx context.Context, extra map[string]interface{}) (map[motor.Feature]bool, error) {
+	return map[motor.Feature]bool{
+		motor.PositionReporting: true,
+	}, nil
+}
+
+// IsMoving returns if the motor is currently moving.
+func (m *uln28byj) IsMoving(ctx context.Context) (bool, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return m.stepPosition != m.targetStepPosition, nil
+}
+
+// Stop turns the power to the motor off immediately, without any gradual step down.
+func (m *uln28byj) Stop(ctx context.Context, extra map[string]interface{}) error {
```

Comment:
the way this is written, this doesn't _immediately_ turn the motor off. 
If what we want is to do that, I think we should get a cancel context when initializing the motor and have this kind of select statement at the beginning of the for-loop in  `doStep`
```
select {
case <-m.cancelCtx.Done():
	return errors.New("context canceled")
default:
}
```
If not, then the comment should be re-worded to say that the motor will stop at the end of the current step in the sequence

Reply:
Tested this out, it doesn't allow motor to re-run once context is canceled.